### ISSUE TITLE
Include language packs in build

### ISSUE THREAD

**Issue by nico151999** [nguyenq/tess4j#84]
This is more a question than an issue. First I tried to compile the library manually without any configuration besides from the default ones shiped with the sources using NetBeans. I found out that the build seemed to run successful. However, running the same code on my build as I did on the precompiled library before caused errors. I couldn't resolve that issue. The main reason why I want to make an own build is that I want to ship a standalone program which includes the deu.traineddata pack. On windows the tessdata folder can be found under C:\Users\testUser\AppData\Roaming\testProject\testProjectTesseract\tessdata. If I save the language pack manually under this directory and if I use the precompiled library from the maven repositories, I can set the deu language in my Java project. However, the deu.traineddata is intended to be created automatically by the library. How can I compile the library so that it runs without errors and does automatically include and create the German (deu) language pack.

I used these sources:
https://sourceforge.net/projects/tess4j/files/tess4j/3.4.3/Tess4J-3.4.3-src.zip/download

This is the language pack:
https://github.com/tesseract-ocr/tessdata/raw/3.04.00/deu.traineddata

Thank you very much for your support!

**Comment by nguyenq** [comment]
Only `eng.traineddata` is bundled in the distro. Therefore, you'll need to manually include your `deu.traineddata` in the `tessdata` folder when you package your distro.

**Comment by nico151999** [comment]
I should have mentioned that I'm forced to work on Windows which does not have any language pack included. The eng.traineddata I access only exist since I'm using SikuliX-API in the same project. SikuliX uses an old version of Tesseract. That's why I would prefer tess4j. However, the computers the final application will run on do not have admin rights. I also mustn't run any installer on it. That's why the jar has to work without any dependencies that are not baked into it. Is there no chance to bake language files into the final jar?

**Comment by nguyenq** [comment]
You can try the version on GitHub. It's Maven based and can include `.traineddata` file in the JAR. Then during execution, you can call `LoadLibs.extractTessResources("tessdata")` to extract `tessdata` folder from the JAR.

http://tess4j.sourceforge.net/tutorial/

**Comment by nico151999** [comment]
Awesome, first I'd like to thank you very much. I compiled it using IntelliJ IDEA and Maven Jar plugin. I've got a packaged jar file now with every resource I want to be inside the Jar. One issue is still left, however: Running

`File imageBase = new File(pathTempImage);`

causes

`File imageBase = new File(pathTempImage);`

Any suggestions? Thanks a lot in advance!

**Comment by nguyenq** [comment]
What error did it cause?

**Comment by nico151999** [comment]
My project's structure is as follows:

```
.idea
--some ide specific configurations
gradle
--wrapper
----the gradle wrapper
src
--main
----java
------Main.java
----resources
------libs
--------tess4j built from master branch with deu.traineddata
--------other libraries required to run tess4j
------pics
--------some PNG files
--test
----irrelevant stuff
build.gradle
gradlew
gradlew.bat
settings.gradle
```

The content of my build.gradle is that:

`import org.apache.tools.ant.taskdefs.condition.Os

buildscript {
    repositories {
        jcenter()
    }
    dependencies {
        classpath 'edu.sc.seis.gradle:launch4j:2.4.3'
    }
}

group 'de.telekom'
version '1.1.1'

apply plugin: 'java'
apply plugin: 'edu.sc.seis.launch4j'

sourceCompatibility = 1.8

ext {

    if (Os.isFamily(Os.FAMILY_UNIX)) {
        os = 'linux'
    }
    if (Os.isFamily(Os.FAMILY_MAC)) {
        os = 'macosx'
    }
    if (Os.isFamily(Os.FAMILY_WINDOWS)) {
        os = 'windows'
    }
    arch = System.getProperty("os.arch")

    if(arch == 'amd64'){
        arch = 'x86_64'
    }
    opencvBinaryClassifer = os+'-'+arch

}

repositories {
    mavenCentral()
    jcenter()
}

dependencies {
    testCompile group: 'junit', name: 'junit', version: '4.12'
}
group 'de.telekom'
version '1.1.1'

apply plugin: 'java'

sourceCompatibility = 1.8

repositories {
    mavenCentral()
}

launch4j {
    mainClassName = 'Main'
    outfile = 'sikuliX.exe'
}

dependencies {
    testCompile group: 'junit', name: 'junit', version: '4.12'
    compile fileTree(dir: 'src/main/resources/libs', include: '*.jar')
    compile ('com.sikulix:sikulixapi:1.1.1') {
        exclude module: '${sikulix.libs}'
        exclude module: 'jadb'
    }
    switch(System.getProperty('os.name').toLowerCase().split()[0]) {
        case 'windows':
            compile group: 'com.sikulix', name: 'sikulixlibswin', version: '1.1.1'
            break
        case 'linux':
            compile group: 'com.sikulix', name: 'sikulixlibslux', version: '1.1.1'
            break
        case 'mac':
            compile group: 'com.sikulix', name: 'sikulixlibsmac', version: '1.1.1'
            break
        default:
            throw new Exception('Unknown OS')
    }
}

jar {
    from {
        configurations.runtime.collect {it.isDirectory() ? it : zipTree(it)}
    }
    manifest {
        attributes("Main-Class": "Main" )
    }
}`

Currently the project runs fine if I use IntelliJ to run the main method. If I use gradle to build a jar and run this jar I'm still facing the unsatisfied link error though. I've watched forum posts of other users of tess4j facing the unsatisfied link error but nothing really helps me to solve the issue. Any ideas?

`Exception in thread "main" java.lang.UnsatisfiedLinkError: The specified module could not be found.

        at com.sun.jna.Native.open(Native Method)
        at com.sun.jna.Native.open(Native.java:1759)
        at com.sun.jna.NativeLibrary.loadLibrary(NativeLibrary.java:260)
        at com.sun.jna.NativeLibrary.getInstance(NativeLibrary.java:398)
        at com.sun.jna.Library$Handler.<init>(Library.java:147)
        at com.sun.jna.Native.loadLibrary(Native.java:412)
        at com.sun.jna.Native.loadLibrary(Native.java:391)
        at net.sourceforge.tess4j.util.LoadLibs.getTessAPIInstance(LoadLibs.java
:81)
        at net.sourceforge.tess4j.TessAPI.<clinit>(TessAPI.java:42)
        at net.sourceforge.tess4j.Tesseract.init(Tesseract.java:364)
        at net.sourceforge.tess4j.Tesseract.doOCR(Tesseract.java:277)
        at net.sourceforge.tess4j.Tesseract.doOCR(Tesseract.java:209)
        at net.sourceforge.tess4j.Tesseract.doOCR(Tesseract.java:193)
        at Main.main(Main.java:137)`

**Comment by nguyenq** [comment]
`UnsatisfiedLinkError` means the app is unable to locate and load the native libraries, which also includes required VC++ runtime. On Windows, Tess4J needs Tesseract and Leptonica DLLs. During runtime, they are extracted from their respective JAR to the user's Temp folder and loaded from there. It's good to start from a clean slate, so you may wan to wipe out `tess4j` folder under `Temp` directory and run the test again. Dependency Walker could be used to determine whether all the necessary DLLs were successfully loaded.

**Comment by nico151999** [comment]
Thank you, I'll try it out tomorrow morning. That still does not explain why it works if I run it from inside IntelliJ. Anyway, thanks for sharing your experience!

**Comment by nico151999** [comment]
I tried out your suggestions. I tried running it on a new Windows host to be 100% sure there are no temp files which might cause errors. The same problem appears again though. I do not know exactly how to use Dependency Walker, especially with Java programs. I packaged an exe file using launch4j and opened it in dependency walker (don't know if that's the right way) and the result wasn't very satisfying. It did not help at all. That's the output...
<img width="840" alt="screen" src="https://user-images.githubusercontent.com/10008855/36592257-3c279eb6-1895-11e8-95cc-b37420758006.PNG">

**Comment by nguyenq** [comment]
Do all the DLLs exist under `c:\temp\tess4j\win32-x86-64\` ?

The `LoadLibs` extracts them to the Temp directory defined by Java's `java.io.tmpdir` property and set `jna.library.path` to that path for JNA to search for the DLLs during runtime. I'm not familiar with gradle and not sure that it would alter the default behavior of JNA. If `jna.library.path` variable was accidentally changed to an incorrect path, `UnsatisfiedLinkError` exceptions would occur.

**Comment by nico151999** [comment]
The 3 DLLs are extracted so they exist under c:\temp\tess4j\win32-x86-64\. I also use LoadLibs class to make sure this is done. As they exist under the mentioned temp dir, I assume java.io.tmpdir is set properly. Since I still got the UnsatisfiedLinkError, I've already tried setting jna.library.path manually. Nothing seems to help and for some reason after the most recent Windows restart running the Program from within IntelliJ does not work any more as well. I made sure it is not caused by Gradle. The same behavior appears using a maven project or using a default IntelliJ project. Using a Windows 7 VM does not help, either. The same behavior, no matter what I do.
It's interesting that 'ITesseract test = new Tesseract();' runs fine. The error appears as soon as the doOCR method is called. I'm a bit disappointed still hoping to get this running.

**Comment by nguyenq** [comment]
The library has hard dependency on VC++ 2015, so make sure you have the appropriate one (32- or 64-bit) installed.

What version of Tess4J are you running? What happen when you run at command prompt: `mvn test`

**Comment by nico151999** [comment]
Yep, VC++ 2015 is installed. Tess4J 4.0.0 built from master branch. I'm at my family this weekend. Will be back at my machine next monday. I'll try mvn test by then. Thank you so much again for your support!

**Comment by nico151999** [comment]
Just a status update:

I've played around a bit. I recreated the project as a maven project as I don't really care if I use Maven or Gradle. I need to learn some things about Maven first but I thought it would be better for applying your advises.  I'm using the tess4j 3 version from the maven repositories right now. It does not work as satisfying as version 4 did with the tesseract_best-language pack until it stopped working for some reason. Well, is it possible to implement tess4j v4 in my maven project? Would be absolutely awesome. Thanks in advance.

BTW, I decided to implement the language pack by copying it to src/main/resources/tessdata/deu.traineddata.
